macros/edr/system/system_utils/buckets_cte.sql

```diff
@@ -0,0 +1,58 @@
+{% macro complete_buckets_cte(time_bucket) %}
+    {{ adapter.dispatch('complete_buckets_cte','elementary')(time_bucket) }}
+{% endmacro %}
+
+{# Databricks and Spark #}
+{% macro default__complete_buckets_cte(time_bucket) %}
+    select
+      edr_bucket_start,
+      {{ elementary.timeadd(time_bucket.period, time_bucket.count, 'edr_bucket_start') }} as edr_bucket_end
+    from (select explode(sequence({{ elementary.cast_as_timestamp(elementary.quote(elementary.get_min_bucket_start())) }}, {{ elementary.cast_as_timestamp(elementary.quote(elementary.get_max_bucket_end())) }}, interval {{ time_bucket.count }} {{ time_bucket.period }})) as edr_bucket_start)
+    where edr_bucket_end < {{ elementary.cast_as_timestamp(elementary.quote(elementary.get_max_bucket_end())) }}
+{% endmacro %}
+
+{% macro snowflake__complete_buckets_cte(time_bucket) -%}
+    {%- set complete_buckets_cte %}
+        with dates as (
+          select {{ elementary.cast_as_timestamp(elementary.quote(elementary.get_min_bucket_start())) }} as edr_bucket_start
+          union all
+          select {{ elementary.timeadd(time_bucket.period, time_bucket.count, 'edr_bucket_start') }} as bucket_end
+          from dates
+          where bucket_end < {{ elementary.cast_as_timestamp(elementary.quote(elementary.get_max_bucket_end())) }}
+        )
+        select
+          edr_bucket_start,
+          {{ elementary.timeadd(time_bucket.period, time_bucket.count, 'edr_bucket_start') }} as edr_bucket_end
+        from dates
+        where edr_bucket_end < {{ elementary.cast_as_timestamp(elementary.quote(elementary.get_max_bucket_end())) }}
+    {%- endset %}
+    {{ return(complete_buckets_cte) }}
+{% endmacro %}
+
+
+{% macro bigquery__complete_buckets_cte(time_bucket) %}
+    {%- set complete_buckets_cte %}
+        select
+          edr_bucket_start,
+          {{ elementary.timeadd(time_bucket.period, time_bucket.count, 'edr_bucket_start') }} as edr_bucket_end
+        from unnest(generate_timestamp_array({{ elementary.cast_as_timestamp(elementary.quote(elementary.get_min_bucket_start())) }}, {{ elementary.cast_as_timestamp(elementary.quote(elementary.get_max_bucket_end())) }}, interval {{ time_bucket.count }} {{ time_bucket.period }})) as edr_bucket_start
+        where edr_bucket_end < {{ elementary.cast_as_timestamp(elementary.quote(elementary.get_max_bucket_end())) }}
+    {%- endset %}
+    {{ return(complete_buckets_cte) }}
+{% endmacro %}
+
+{% macro redshift__complete_buckets_cte(time_bucket) %}
+    {%- set complete_buckets_cte %}
+      with integers as (
+        select (row_number() over (order by 1)) - 1 as num
+        from pg_catalog.pg_class
```

Comment:
Hh what is this sorcery? ;)
The trick is nice but what happens if pg_class doesn't have enough rows?

Reply:
It would be problematic.
However, I looked up various [system catalogs](https://www.postgresql.org/docs/current/catalogs.html) and this one seemed to have the biggest row count.

Reply:
Can we maybe add a simple check and log a warning if that's the case? (seems like it should be rare so maybe it's good enough, but just so we'll know)

Reply:
Another concern is that a user might not have access to query the `pg_catalog` schema but I think that's rather unlikely. Kind of absurd that Redshift doesn't have an easy way of doing `range(start, end)`.

Reply:
sababa we can start with that and learn
another thing you could do I guess is to create a temp table which has the range, but maybe overkill